xonsh/completer.py

```diff
@@ -396,27 +418,30 @@ def attr_complete(self, prefix, ctx):
         return attrs
 
     def _all_commands(self):
+        """Possibly update our set of cached commands and return the set."""
         path = builtins.__xonsh_env__.get('PATH', [])
-        # did PATH change?
+        # Did PATH change?
         path_hash = hash(tuple(path))
         cache_valid = path_hash == self._path_checksum
         self._path_checksum = path_hash
-        # did aliases change?
+        # Did aliases change?
         al_hash = hash(tuple(sorted(builtins.aliases.keys())))
         self._alias_checksum = al_hash
         cache_valid = cache_valid and al_hash == self._alias_checksum
         pm = self._path_mtime
-        # did the contents of any directory in PATH change?
-        for d in filter(os.path.isdir, path):
+        # Did the contents of any directory in PATH change?
+        for d in (p for p in path if os.path.isdir(p)):
```

Comment:
The functional idioms such as map and filter are appealing, especially to anyone familiar with a language like Scala or Lisp, but antithetical to the Python way of doing things. Which is why pylint complains about their use.